### DISCUSSION TITLE
Child Lock on ClickSmart Zigbee Wall Outlet (Tuya TS011F)

### DISCUSSION THREAD

**Discussion by simonaldrich** [arendst/Tasmota#14727] [Q&A] [answered]
Hi All, I originally posted this question to [r/Tasmota](https://www.reddit.com/r/tasmota/comments/sjy5ou/zigbee_command_to_lock_power_buttons/) but I've not had any luck there so I'm hoping someone here might have some ideas.

I'm using Tasmota on the Sonoff Zigbee bridge and it's working flawlessly. However, I recently got a new device which has some functionality I'd like to use and I can't figure out what Zigbee commands to send.

The device I've got is a [ClickSmart 2-gang wall outlet](https://zigbee.blakadder.com/ClickSmart_CMA30036.html). It looks like it's based on a Tuya TS011F chip. I can control the power for both sockets (endpoints 1 & 2) just fine but according to the [manufacturer](https://click-smart.com/products/mode-smart-sockets) each outlet can be "locked On or Off using the app". The manual says that "Pressing the power button when sockets are in "lock" mode will result in the status LED flashing x3 times". I'd really like to be able to use this functionality as it would be super-useful with small children in the house.

Does anyone have any ideas how I might go about working out what to [ZbSend](https://tasmota.github.io/docs/Zigbee/#low-level-commands) to the device to use this "locking" functionality?

The ZbStatus3 response for the device looks like:
```
stat/zigbee/RESULT = {"ZbStatus3":[{"Device":"0x5F10","IEEEAddr":"0x84FD27FFFECA551E","ModelId":"TS011F","Manufacturer":"_TYZB01_hlla45kx","Endpoints":[1,2],"Config":["O01","O02"],"Power":0,"Power2":0,"Reachable":true,"LastSeen":22,"LastSeenEpoch":1643924707,"LinkQuality":120}]}
```
On powering-on the device I get the following through on MQTT:
```
tele/zigbee/SENSOR {"ZbReceived":{"0x5F10":{"Device":"0x5F10","Power":0,"0006/4001":0,"0006/4002":0,"Endpoint":1,"LinkQuality":107}}}
tele/zigbee/SENSOR {"ZbReceived":{"0x5F10":{"Device":"0x5F10","Power":0,"0006/4001":0,"0006/4002":0,"Endpoint":2,"LinkQuality":102}}}
tele/zigbee/SENSOR {"ZbReceived":{"0x5F10":{"Device":"0x5F10","Power":0,"Endpoint":1,"LinkQuality":84}}}
tele/zigbee/SENSOR {"ZbReceived":{"0x5F10":{"Device":"0x5F10","AppVersion":null,"0000/FFE2":22,"Endpoint":1,"LinkQuality":105}}}
tele/zigbee/SENSOR {"ZbReceived":{"0x5F10":{"Device":"0x5F10","AppVersion":64,"0000/FFE2":22,"Endpoint":1,"LinkQuality":107}}}
tele/zigbee/SENSOR {"ZbReceived":{"0x5F10":{"Device":"0x5F10","0000/FFDF":"","Endpoint":1,"LinkQuality":107}}}
tele/zigbee/SENSOR {"ZbReceived":{"0x5F10":{"Device":"0x5F10","Cluster":10,"Read":[7],"ReadNames":{"LocalTime":true},"Endpoint":1,"LinkQuality":94}}}
```

I originally wondered if the `0006/4001` (`OnTime`) and `0006/4002` (`OffWaitTime`) attributes (according to the [Zigbee Cluster Library Specification](https://zigbeealliance.org/wp-content/uploads/2019/12/07-5123-06-zigbee-cluster-library-specification.pdf)) might have something to do with the locking but it turns out they're used for the [countdown feature](https://developer.tuya.com/en/docs/iot-device-dev/tuya-zigbee-measuring-smart-plug-access-standard?id=K9ik6zvofpzqk#title-12-Countdown). Setting them can be done with command 0x42 of Cluster 6. Here's an example of setting the countdown feature for Endpoint 1 to 10 seconds:
```
cmnd/zigbee/ZbSend {"Device":"0x5F10", "Endpoint": 1, "Send":"0006!42/010A000A00"}
stat/zigbee/RESULT {"ZbSend":"Done"}
tele/zigbee/SENSOR {"ZbReceived":{"0x5F10":{"Device":"0x5F10","0006/4002":10,"Endpoint":1,"LinkQuality":126}}}
tele/zigbee/SENSOR {"ZbReceived":{"0x5F10":{"Device":"0x5F10","Power":0,"0006/4001":0,"0006/4002":0,"Endpoint":1,"LinkQuality":123}}}
cmnd/zigbee/ZbSend {"Device":"0x5F10", "Endpoint": 1, "Send":"0006!42/010A000A00"}
stat/zigbee/RESULT {"ZbSend":"Done"}
tele/zigbee/SENSOR {"ZbReceived":{"0x5F10":{"Device":"0x5F10","0006/4001":10,"Endpoint":1,"LinkQuality":128}}}
tele/zigbee/SENSOR {"ZbReceived":{"0x5F10":{"Device":"0x5F10","Power":1,"0006/4001":0,"0006/4002":0,"Endpoint":1,"LinkQuality":128}}}
```
When the countdown timer has expired the power state of the endpoint is toggled.

I did find some references to a Tuya specific Zigbee cluster ID of [0xEF00](https://developer.tuya.com/en/docs/iot-device-dev/tuya-zigbee-universal-docking-access-standard?id=K9ik6zvofpzql#title-3-Endpoint) which I notice that Tasmota already has some [support](https://github.com/arendst/Tasmota/blob/development/tasmota/xdrv_23_zigbee_5_converters.ino#L630) for. Disappointingly that didn't magically work and seemed to indicate that Cluster ID (`0xEF00`) isn't supported by the device:
```
cmnd/zigbee/ZbSend {"Device":"0x5F10", "Endpoint":1, "Write":{"TuyaChildLock":1}}
stat/zigbee/RESULT {"ZbSend":"Done"}
tele/zigbee/RESULT {"ZbResponse":{"Device":"0x5F10","Command":"EF00!00","Status":195,"StatusMessage":"UNSUPPORTED_CLUSTER","Endpoint":1,"LinkQuality":110}}
```
Not sure where to go from here but I'm hoping you fine folks might. Any and all ideas and suggestions gratefully accepted :slightly_smiling_face:

**Comment by simonaldrich** [comment, marked as answer]
I bit the bullet and bought a Tuya Zigbee Hub and a CC2531 stick and sniffed the traffic. I found that the child-lock functionality is (de)activated by setting or clearing a boolean attribute `0x8000` in cluster `0x0006`:
```
No.     Time           Source                Destination           Protocol Length Info
     11 2.630120       0x0000                0xf47c                ZigBee HA 52     ZCL: Write Attributes, Seq: 108

Frame 11: 52 bytes on wire (416 bits), 52 bytes captured (416 bits) on interface -, id 0
IEEE 802.15.4 Data, Dst: 0xf47c, Src: 0x0000
ZigBee Network Layer Data, Dst: 0xf47c, Src: 0x0000
ZigBee Application Support Layer Data, Dst Endpt: 1, Src Endpt: 1
    Frame Control Field: Data (0x00)
    Destination Endpoint: 1
    Cluster: On/Off (0x0006)
    Profile: Home Automation (0x0104)
    Source Endpoint: 1
    Counter: 32
ZigBee Cluster Library Frame, Command: Write Attributes, Seq: 108
    Frame Control Field: Profile-wide (0x10)
        .... ..00 = Frame Type: Profile-wide (0x0)
        .... .0.. = Manufacturer Specific: False
        .... 0... = Direction: Client to Server
        ...1 .... = Disable Default Response: True
    Sequence Number: 108
    Command: Write Attributes (0x02)
    Attribute Field, Boolean: 0x01
        Attribute: Unknown (0x8000)
        Data Type: Boolean (0x10)
        0000 0001 = Boolean: True
```
Which means you can use the following `ZbSend` to enable the child-lock:
```
cmnd/zigbee/ZbSend {"Device":"0x5F10", "Endpoint":1, "Write":{"0006/8000%10":1}}
```
and disable it with:
```
cmnd/zigbee/ZbSend {"Device":"0x5F10", "Endpoint":1, "Write":{"0006/8000%10":0}}
```

You can retrieve the lock state by reading the same attribute:
```
cmnd/zigbee/ZbSend {"Device":"0x5F10", "Endpoint":1, "Send":"0600_00/0080"}
```
which ultimate returns something like this to the console:
```
ZIG: ZbZCLRawReceived: {"0x5F10":{"0006/8000":1,"Endpoint":1,"LinkQuality":134}}
```

**Reply by s-hadinger** [reply]
Good finding. There is a simpler way to read an attribute value than `{"Device":"0x5F10", "Endpoint":1, "Send":"0600_00/0080"}`

You can try
```
cmnd/zigbee/ZbSend {"Device":"0x5F10", "Endpoint":1, "cluster":"0x0006", "read":"0x8000"}
```
